### PR TITLE
Slack corporate import

### PR DESCRIPTION
**Pull request by svelle**
I got my hands on a current export and was able to fix a lot of issues that we had in the current state.
It's almost 3am here and I've been working on this for nearly 12 hours straight so there might be some parts that could be improved :D So I'll mark this as a WIP for now. 
I hope I can straighten any eventual issues out tomorrow.

Also yay to slack for inconsistent channel models in their import files :)

### CODE REVIEW

**Comment by svelle** [comment]
Customer got back to me and it seems to have worked with their dataset, including attachments and image previews.

File: app/slackimport.go [hunk 1]
```diff
@@ -109,6 +114,7 @@ func SlackParseChannels(data io.Reader, channelType string) ([]SlackChannel, err
 
 	var channels []SlackChannel
 	if err := decoder.Decode(&channels); err != nil {
+		// TODO: investigate
```
Comment (lieut-data) [inline app/slackimport.go]: Do we need this in the code? Maybe a ticket instead?

File: app/slackimport.go [hunk 2]
```diff
@@ -475,6 +481,7 @@ func (a *App) addSlackUsersToChannel(members []string, users map[string]*model.U
 
 func SlackSanitiseChannelProperties(channel model.Channel) model.Channel {
 	if utf8.RuneCountInString(channel.DisplayName) > model.CHANNEL_DISPLAY_NAME_MAX_RUNES {
+		// TODO: Investigate
```
Comment (lieut-data) [inline app/slackimport.go]: Same question as above.

File: app/slackimport.go [hunk 3]
```diff
@@ -509,9 +516,15 @@ func (a *App) SlackAddChannels(teamId string, slackchannels []SlackChannel, post
 			Type:        sChannel.Type,
 			DisplayName: sChannel.Name,
 			Name:        SlackConvertChannelName(sChannel.Name, sChannel.Id),
-			Purpose:     sChannel.Purpose["value"],
-			Header:      sChannel.Topic["value"],
+			Purpose:     sChannel.Purpose.Value,
+			Header:      sChannel.Topic.Value,
+		}
+
+		// needed to circumvent slacks model for dm chats, or else the messages won't get imported
```
Comment (lieut-data) [inline app/slackimport.go]: nit: `circumvent` has a very different connotation than I expect is meant here. Perhaps you could explain slacks model instead and thus why we use the `Id` as the `Name`?

File: app/slackimport.go [hunk 4]
```diff
@@ -814,7 +829,40 @@ func (a *App) OldImportUser(team *model.Team, user *model.User) *model.User {
...
+func (a *App) OldImportChannel(channel *model.Channel, sChannel SlackChannel, users map[string]*model.User) *model.Channel {
+	if channel.Type == model.CHANNEL_DIRECT {
+		sc, err := a.createDirectChannel(users[sChannel.Members[0]].Id, users[sChannel.Members[1]].Id)
+		if err != nil {
+			return nil
+		}
+
+		return sc
+	}
+	if channel.Type == model.CHANNEL_GROUP {
```
Comment (lieut-data) [inline app/slackimport.go]: nit: let's add a newline before this

File: app/slackimport.go [hunk 5]
```diff
@@ -814,7 +829,40 @@ func (a *App) OldImportUser(team *model.Team, user *model.User) *model.User {
...
+		members := make([]string, len(sChannel.Members))
+		if len(members) < 8 {
```
Comment (lieut-data) [inline app/slackimport.go]: nit: it's a little roundabout to make an array from another value and then /check/ the length of that array as a precondition. Can we just check `len(sChannel.Members)` first instead?

File: app/slackimport.go [hunk 6]
```diff
@@ -527,15 +540,17 @@ func (a *App) SlackAddChannels(teamId string, slackchannels []SlackChannel, post
...
+			mChannel = a.OldImportChannel(&newChannel, sChannel, users)
 			if mChannel == nil {
 				mlog.Warn(fmt.Sprintf("Slack Import: Unable to import Slack channel: %s.", newChannel.DisplayName))
 				importerLog.WriteString(utils.T("api.slackimport.slack_add_channels.import_failed", map[string]interface{}{"DisplayName": newChannel.DisplayName}))
 				continue
 			}
 		}
 
-		a.addSlackUsersToChannel(sChannel.Members, users, mChannel, importerLog)
+		if sChannel.Type == model.CHANNEL_OPEN || sChannel.Type == model.CHANNEL_PRIVATE {
```
Comment (lieut-data) [inline app/slackimport.go]: Can you add a comment here explaining that the members for direct message and group channels are added as part of `OldImportChannel` above?

File: app/slackimport.go [hunk 7]
```diff
@@ -27,12 +27,17 @@ import (
...
 	Members []string          `json:"members"`
-	Topic   map[string]string `json:"topic"`
-	Purpose map[string]string `json:"purpose"`
+	Purpose SlackChannelSub   `json:"purpose"`
+	Topic   SlackChannelSub   `json:"topic"`
 	Type    string
 }
 
+type SlackChannelSub struct {
+	Value	string	`json:value`
```
[the remark this answers is not in the document]
Comment (svelle) [reply app/slackimport.go]: You're right. Fixed in next commit.

File: app/slackimport.go (hunk 1 above)
Comment (svelle) [reply app/slackimport.go]: This is a leftover from the session where I was investigating why some channels wouldn't import. Removed it.

File: app/slackimport.go (hunk 2 above)
Comment (svelle) [reply app/slackimport.go]: Same as above.

File: app/slackimport.go (hunk 3 above)
Comment (svelle) [reply app/slackimport.go]: Have reworded the comment to be more precise. The problem was that the way the importer worked for all the other channels was that it matched the messages to the channel names, but DM Channels apparently don't have a name like the other types do. So setting the `Id` as the `Name` here works around this by giving the direct messages a channel name to match against so they get imported too.

File: app/slackimport.go (hunk 4 above)
Comment (svelle) [reply app/slackimport.go]: fixed.

File: app/slackimport.go (hunk 5 above)
Comment (svelle) [reply app/slackimport.go]: Yes very clean solution, fixed.

File: app/slackimport.go (hunk 6 above)
Comment (svelle) [reply app/slackimport.go]: Added a comment to clarify why the `if` is there.

**Comment by svelle** [comment]
Thanks for the review @lieut-data , I have updated everything accordingly. 
Let me know if you find anything else! :)

File: app/slackimport.go [hunk 8]
```diff
@@ -814,7 +828,40 @@ func (a *App) OldImportUser(team *model.Team, user *model.User) *model.User {
...
+	if channel.Type == model.CHANNEL_DIRECT {
+		sc, err := a.createDirectChannel(users[sChannel.Members[0]].Id, users[sChannel.Members[1]].Id)
+		if err != nil {
+			return nil
+		}
+
+		return sc
+	}
+
+	// check if direct channel has less than 8 members and if not import as public channel instead
```
Comment (grundleborg) [inline app/slackimport.go]:
```suggestion
	// check if direct channel has less than 8 members and if not import as private channel instead
```

File: app/slackimport.go [hunk 9]
```diff
@@ -509,9 +514,15 @@ func (a *App) SlackAddChannels(teamId string, slackchannels []SlackChannel, post
 			Type:        sChannel.Type,
 			DisplayName: sChannel.Name,
 			Name:        SlackConvertChannelName(sChannel.Name, sChannel.Id),
-			Purpose:     sChannel.Purpose["value"],
-			Header:      sChannel.Topic["value"],
+			Purpose:     sChannel.Purpose.Value,
+			Header:      sChannel.Topic.Value,
+		}
+
+		// Direct message channels don't have a name so we set the id as name or else the messages won't get imported.
```
Comment (grundleborg) [inline app/slackimport.go]:
```suggestion
		// Direct message channels in Slack don't have a name so we set the id as name or else the messages won't get imported.
```

**Comment by lindalumitchell** [comment]
Removed QA label; QA testing not needed pre-merge.

**Comment by grundleborg** [comment]
@svelle looks good now. Can you please merge master into this branch and also get the unit tests passing, and then we're good to merge.

**Comment by svelle** [comment]
@grundleborg should be good to merge now. :)

**Pull request merged**